File: music-api/src/main/java/com/music/sale/config/SecurityConfig.java
package com.music.sale.config;

import com.music.sale.infrastructure.security.jwt.JwtAuthenticationFilter;
import com.music.sale.infrastructure.security.oauth2.OAuth2UserLoader;
import com.music.sale.infrastructure.security.oauth2.handler.OAuth2LoginFailureHandler;
import com.music.sale.infrastructure.security.oauth2.handler.OAuth2LoginSuccessHandler;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

@Configuration
@EnableWebSecurity
@RequiredArgsConstructor
public class SecurityConfig {

  private final JwtAuthenticationFilter jwtAuthenticationFilter;

  @Autowired(required = false)
  private ClientRegistrationRepository clientRegistrationRepository;

  @Autowired(required = false)
  private OAuth2UserLoader oAuth2UserLoader;

  @Autowired(required = false)
  private OAuth2LoginSuccessHandler oAuth2LoginSuccessHandler;

  @Autowired(required = false)
  private OAuth2LoginFailureHandler oAuth2LoginFailureHandler;

  @Bean
  public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
    http
        // CSRF 비활성화 (JWT 사용)
        .csrf(AbstractHttpConfigurer::disable)

        // CORS 설정
        .cors(cors -> cors.configure(http))

        // 세션 사용 안 함 (JWT 사용)
        .sessionManagement(
            session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))

        // 요청 권한 설정
        .authorizeHttpRequests(
            auth ->
                auth
                    // Public endpoints
                    .requestMatchers(
                        "/swagger-ui/**", // Swagger UI
                        "/swagger-auth.html", // OAuth2 인증 후 리다이렉트 페이지
                        "/v3/api-docs/**", // API 문서
                        "/api-docs/**", // API 문서 (커스텀 경로)
                        "/swagger-resources/**", // Swagger 리소스
                        "/webjars/**", // Webjars
                        "/h2-console/**", // H2 콘솔
                        "/actuator/**", // Actuator
                        "/api/v1/auth/**",
                        "/api/v1/products/**",
                        "/api/v1/health",
                        "/error")
                    .permitAll()
                    // OAuth2 endpoints
                    .requestMatchers("/oauth2/**", "/login/oauth2/**")
                    .permitAll()
                    // Authenticated endpoints
                    .anyRequest()
                    .authenticated())

        // JWT 필터 추가
        .addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter.class);

    // OAuth2 로그인 설정 (OAuth2 클라이언트가 등록되어 있을 때만 활성화)
    if (clientRegistrationRepository != null && oAuth2UserLoader != null) {
      http.oauth2Login(
          oauth2 ->
              oauth2
                  .userInfoEndpoint(userInfo -> userInfo.userService(oAuth2UserLoader))
                  .successHandler(oAuth2LoginSuccessHandler)
                  .failureHandler(oAuth2LoginFailureHandler));
    }

    return http.build();
  }
}

File: src/main/kotlin/com/music/sale/config/SecurityConfig.kt
// Copyright (C) 2024 Your Name or Company
package com.music.sale.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest
import org.springframework.security.oauth2.client.userinfo.OAuth2UserService
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.AuthenticationSuccessHandler
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
class SecurityConfig(private val jwtAuthenticationFilter: JwtAuthenticationFilter) {
    @Bean
    fun passwordEncoder(): PasswordEncoder {
        return BCryptPasswordEncoder()
    }

    @Bean
    fun corsConfigurationSource(): CorsConfigurationSource {
        val configuration = CorsConfiguration()
        configuration.allowedOrigins = listOf("*")
        configuration.allowedMethods = listOf("GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS")
        configuration.allowedHeaders = listOf("*")
        configuration.allowCredentials = false
        configuration.maxAge = 3600L

        val source = UrlBasedCorsConfigurationSource()
        source.registerCorsConfiguration("/**", configuration)
        return source
    }

    @Bean
    fun oAuth2AuthenticationSuccessHandler(jwtService: com.music.sale.application.auth.service.JwtService): AuthenticationSuccessHandler {
        return com.music.sale.config.oauth2.OAuth2AuthenticationSuccessHandler(jwtService)
    }

    @Bean
    fun customOAuth2UserService(): OAuth2UserService<OAuth2UserRequest, OAuth2User> {
        return com.music.sale.config.oauth2.CustomOAuth2UserService()
    }

    @Bean
    fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .csrf { it.disable() }
            .cors { it.configurationSource(corsConfigurationSource()) }
            .sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
            .addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter::class.java)
            .authorizeHttpRequests { authorize ->
                authorize
                    .anyRequest().permitAll() // 테스트를 위해 모든 요청 허용
            }
            .headers { headers ->
                headers.frameOptions { it.disable() }
            }
            .oauth2Login { oauth2 ->
                oauth2
                    .userInfoEndpoint { userInfo ->
                        userInfo.userService(customOAuth2UserService())
                    }
                    .successHandler(oAuth2AuthenticationSuccessHandler(com.music.sale.application.auth.service.JwtService()))
            }

        return http.build()
    }
}
